src/vpn-api/utils/openvpn.ts

```diff
@@ -0,0 +1,308 @@
+		private mgtPort: number,
+		private subnet: Netmask,
+		private gateway?: string,
+	) {
+		super();
+
+		this.connector = new VpnConnector();
+		// proxy `data` events from connector, splitting at newlines
+		this.connector.connection.on('data', data => {
+			const lines = ((this.buf || '') + data.toString()).split('\r\n');
+			this.buf = lines.pop();
+			for (const line of lines) {
+				this.emit('manager:data', line.trim());
+			}
+		});
+		// subscribe to our own raw data events in order to generate structured events
+		this.on('manager:data', this.dataHandler);
+	}
+
+	private args() {
+		const gateway = this.gateway || this.subnet.first;
+		return [
+			'--status',
+			`/run/openvpn/server-${this.instanceId}.status`,
+			'10',
+			'--cd',
+			'/etc/openvpn',
+			'--config',
+			'/etc/openvpn/server.conf',
+			'--verb',
+			'3',
+			'--dev',
+			`tun${this.instanceId}`,
+			'--port',
+			`${this.vpnPort}`,
+			'--management',
+			'127.0.0.1',
+			`${this.mgtPort}`,
+			'--management-hold',
+			'--ifconfig',
+			gateway,
+			this.subnet.second,
+			'--ifconfig-pool',
+			this.subnet.third,
+			this.subnet.last,
+			'--route',
+			this.subnet.base,
+			this.subnet.mask,
+			'--push',
+			`route ${gateway}`,
+			'--management-client-auth',
+		];
+	}
+
+	private dataHandler = (data: string) => {
+		if (data.substr(0, 5) === '>LOG:') {
+			// >LOG:{timestamp},{level},{message}
+			const logData = data.substr(5);
+			const [, level, ...message] = logData.split(',');
+			this.emit(
+				'log',
+				VpnLogLevels[(level as keyof typeof VpnLogLevels) || 'n'],
+				message.join(','),
+			);
+		} else if (data.substr(0, 8) === '>CLIENT:') {
+			// >CLIENT:{event},{clientId},{...args}
+			// >CLIENT:ENV,{key}={value}
+			// >CLIENT:END
+			for (const eventType of [
+				'connect',
+				'address',
+				'established',
+				'disconnect',
+			]) {
+				if (data.substr(8, eventType.length) === eventType.toUpperCase()) {
+					const [clientId, ...eventArgs] = data
+						.substr(8 + eventType.length + 1)
+						.split(',');
+					this.off('manager:data', this.dataHandler);
+					this.on(
+						'manager:data',
+						this.clientEventEmitterFactory(
+							eventType,
+							parseInt(clientId, 10),
+							eventArgs,
+						),
+					);
+				}
+			}
+		} else if (data.substr(0, 15) === '>BYTECOUNT_CLI:') {
+			// >BYTECOUNT_CLI:{clientId},{bytesRx},{bytesTx}
+			const [clientId, bytesReceived, bytesSent] = data.substr(15).split(',');
+			this.emit('client:bytecount', parseInt(clientId, 10), {
+				bytes_received: parseInt(bytesReceived, 10),
+				bytes_sent: parseInt(bytesSent, 10),
+			});
+		}
+	};
+
+	private clientEventEmitterFactory = (
+		eventName: string,
+		clientId: number,
+		eventArgs?: string[],
+	) => {
+		const env: { [key: string]: number | string } = {};
+		const emitter = (data: string) => {
+			if (data.substr(1, 10) === 'CLIENT:ENV' && data.substr(12) !== 'END') {
+				// >CLIENT:ENV,key=val
+				const msg = data.substr(12);
```

Comment:
You mean like `const msg = data.substr('>CLIENT:ENV'.length);`?

Reply:
Yeah, that way it avoids the potential for things to fall out of sync.  Also because it's a common pattern here it might be worth adding a helper along the lines of
```ts
const x = (data: string, msg: string) => {
	if (data.startsWith(msg)) {
		return data.substr(msg.length);
	}
	return false
}
```

Reply:
The only issue is because of needing to keep the result but also check it then the `if` statements can get a little awkward

Reply:
Yeah I was just about to ask about that as your second comment landed!